controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java

```diff
@@ -76,6 +80,106 @@
     private static final OpStatsLogger DELETE_STREAM = STATS_LOGGER.createStats(MetricsNames.DELETE_STREAM);
     private final static String RESOURCE_PART_SEPARATOR = "_%_";
 
+    @Data
+    /**
+     * This class represents a 96 bit number with 32 bit msb encoded as integer and 64 bit lsb encoded as long.
+     * It is unsigned and only allows for non negative values. It also implements comparable interface and comparison involves
+     * first compariging msbs and if msbs are equal then we compare lsbs.
+     */
+    protected static class BigLong implements Comparable {
+        public static final BigLong ZERO = new BigLong(0, 0L);
+        private final int msb;
+        private final long lsb;
+
+        BigLong(int msb, long lsb) {
+            Preconditions.checkArgument(msb >= 0);
+            Preconditions.checkArgument(lsb >= 0);
+
+            this.msb = msb;
+            this.lsb = lsb;
+        }
+
+        @Override
+        public int compareTo(Object o) {
+            if (!(o instanceof BigLong)) {
+                throw new RuntimeException("incomparable objects");
+            }
+            BigLong other = (BigLong) o;
+
+            if (msb != other.msb) {
+                return Integer.compare(msb, other.msb);
+            } else {
+                return Long.compare(lsb, other.lsb);
+            }
+        }
+
+        byte[] toBytes() {
+            byte[] b = new byte[Integer.BYTES + Long.BYTES];
+            BitConverter.writeInt(b, 0, msb);
+            BitConverter.writeLong(b, Integer.BYTES, lsb);
+            return b;
+        }
+
+        static BigLong fromBytes(byte[] b) {
+            int msb = BitConverter.readInt(b, 0);
+            long lsb = BitConverter.readLong(b, Integer.BYTES);
+
+            return new BigLong(msb, lsb);
+        }
+
+        static BigLong add(BigLong counter, int increment) {
+            BigLong retVal;
+            if (counter.lsb < Long.MAX_VALUE - increment) {
+                retVal = new BigLong(counter.msb, counter.lsb + increment);
+            } else if (counter.msb < Integer.MAX_VALUE) {
+                int remainder = increment - (int) (Long.MAX_VALUE - counter.lsb);
+                retVal = new BigLong(counter.msb + 1, remainder);
+            } else {
+                // overflow: throw exception
+                throw new ArithmeticException("Overflow");
+            }
+
+            return retVal;
+        }
+    }
+
+    /**
+     * This class provides the ability to atomically update a BigLong value.
+     */
+    protected static class AtomicBigLong {
+        @GuardedBy("lock")
+        private BigLong value;
```

Comment:
Use `AtomicReference`. It's more lightweight.

Reply:
we use lock as opposed to atomic reference because we want to give the semantics of "incrementAndGet" without conflicts. With AtomicReference we will get compareAndSet and that can fail with concurrent increments as first will succeed while others will have to retry. using a lock here makes it cleaner and more efficient.